### DISCUSSION TITLE
field names for the PDF templates

### DISCUSSION THREAD

**Discussion by Ilamparithi-Tamizhanangu** [YetiForceCompany/YetiForceCRM#15585] [How to] [answered]
Hi,

As continuation with #9840 discussion, we are having the below syntax for creating the tables in the quotation template. 

$(inventory : type=table columns=name,qty,price href=yes)$
$(inventory : type=table columns=seq,name,qty,unit,price,total,net href=no)$

It will create the table as below. It will create the table in the quotation with predefined column names.

![image](https://user-images.githubusercontent.com/96602176/147423403-8c33b57a-1c28-47b6-8e1a-ef57563ded1a.png)

This is the product column in the quotation.

![image](https://user-images.githubusercontent.com/96602176/147423425-6c2733c3-9633-4af2-a173-fc16a0f62456.png)

Our queries are,
1.  How to adjust the column width for each column and give different names for the row headers (Like we wish to use Qty instead of Quantity).
2.  Also i wish to get the product description alone. We dont want to get the item name. How to get only comment in the products table. (It works in the dynamic inventory columns. 
3. Please post the keywords for selecting the all the columns available in the products table.
4. How to adjust the field width of each column in the quotation template. 

Thanks in advance.

with regards,
Rajasekaran.

**Comment by mariuszkrzaczkowski** [comment]
> 3. Please post the keywords for selecting the all the columns available in the products table.

each inventory module has a table `...._invfield` example: `u_yf_squotes_invfield`
![image](https://user-images.githubusercontent.com/10198654/147445060-b2d1b15e-9433-47e3-afbb-6e591a3ea603.png)

**Reply by mariuszkrzaczkowski** [reply]
> 2. Also i wish to get the product description alone. We dont want to get the item name. How to get only comment in the products table. (It works in the dynamic inventory columns.

comment1

**Reply by mariuszkrzaczkowski** [reply]
> 1. How to adjust the column width for each column and give different names for the row headers (Like we wish to use Qty instead of Quantity).

>4. How to adjust the field width of each column in the quotation template.

this is not currently possible, but it is easy to do

**Reply by Ilamparithi-Tamizhanangu** [reply]
hi,

Thanks for the help. Now we can able to generate the table with required columns with the keywords.

Please find the below template. 
![image](https://user-images.githubusercontent.com/96602176/147457808-08642e18-2195-42a5-85f5-e3b5a123b5b2.png)

This is the dynamic table field values. 
![dynamic template](https://user-images.githubusercontent.com/96602176/147457846-3aa99b06-78e2-4e9f-96cd-7f268006b557.png)

The output is, 
![output](https://user-images.githubusercontent.com/96602176/147457868-85746b3b-2485-4358-b507-38ce10b3da4d.png)

The last key word is not visible in the PDF (gross value in the above screenshot), we tried with different keywords. Always, the last column was not displayed in the PDF output.

Please guide to solve this.

Thanks.

**Reply by Ilamparithi-Tamizhanangu** [reply]
> > 1. How to adjust the column width for each column and give different names for the row headers (Like we wish to use Qty instead of Quantity).
> 
> > 4. How to adjust the field width of each column in the quotation template.
> 
> this is not currently possible, but it is easy to do

hi,

If possible, could you please share the procedure to achieve the same. ( we wish to replace the comment as Product Descripton, Quantity as Qty). 

Thanks.

**Reply by Ilamparithi-Tamizhanangu** [reply]
hi @mariuszkrzaczkowski,

One of the main feature to adapt CRM is quotation generation and maintaining its updates. Creating the PDF output file is one of the primary usage for us. So we need to modify the row headers to generate the PDF file for quotes. 

If we can able to implement this functionalities, please share the procedure or technical materials to do the same. It will help us a lot to generate the PDF from the quotation module.

Thanks & Regards,
Rajasekaran.

**Reply by vovpff** [reply]
You can code text parser function what will render products table as you need. Usually you need to code it once.

**Reply by Ilamparithi-Tamizhanangu** [reply]
Hi, 

I am new to yetiforce crm. earlier i was using Vtiger CRM and we edit the source file for the table alignments. 

I am unable to find the table formating options in the source section of the quote template. 
![image](https://user-images.githubusercontent.com/96602176/147932309-e543cf45-4219-4736-a818-c91504a80926.png)

Could you please help me to locate the file and share the syntax or one example to edit text parser function to prepare the table in the quotation module.

Thanks in advance. 

Thanks & Regards,
Rajasekaran.

**Reply by vovpff** [reply]
Check files for priducts table at /app/TextParser/

**Reply by Ilamparithi-Tamizhanangu** [reply]
Hi,

Thanks for the answer.

I can able to find the product tables with different names used in the template file (like table log version, table short version) which products different PDF outputs. 

Now we are using "$(inventory : type=table columns=seq,comment1,qty,price,total,discount,net,tax,gross href=yes)$" to select the columns for the PDF output. 

Could you please help me to identify the correct file to be edited and any example to change the formating of the row headers in the table output. 

Thanks & Regards,
Rajasekaran

**Reply by vovpff** [reply]
app/TextParser/DynamicInventoryColumnsTable.php

I think this one

**Reply by Ilamparithi-Tamizhanangu** [reply]
Hi,

 I tried to take the copy of existing text parser file (Thanks to #11491) and modified the column list based on the requirement. 

I am having two queries. 
1. How to display two fields in one column of the table. ( like merging the qty (6) and unit (bundle) fields to one colum ( 6 bundle))
2. I am unable to find the example to edit for the row headers of the table. Please share the example to modify the row headers.

Thanks in advance.

Thanks & regards,
Rajasekaran.

**Reply by vovpff** [reply]
Check step by step how to the table code generated and duplicate it.

**Reply by Ilamparithi-Tamizhanangu** [reply]
hi,

we need to change the table headers in the below code. 

`$html .= '<table class="products-table-long-version" style="width:100%;font-size:8px;border-collapse:collapse;">
				<thead>
					<tr>'; 
		$groupModels = [];
		foreach (['ItemNumber', 'Name',  'Quantity', 'UnitPrice', 'TotalPrice', 'Tax', 'GrossPrice']  as $fieldType) {
			foreach ($inventory->getFieldsByType($fieldType) as $fieldModel) {
				$columnName = $fieldModel->getColumnName();
				if (!$fieldModel->isVisible()) {
					continue;
				}
				$html .= "<th style=\"{$headerStyle}\">" . \App\Language::translate($fieldModel->get('label'), $this->textParser->moduleName) . '</th>';
				$groupModels[$columnName] = $fieldModel; 
			}
		} 
		
		
		$html .= '</tr></thead>';`

It uses the table header as label. Instead of using default values, Could you please help me to insert the custom table headers for the products table. 

Thanks & regards,
Rajasekaran.

**Reply by Ilamparithi-Tamizhanangu** [reply]
hi,

Could you please help me to update the custom text (array of the text variables) to the column headers in the below variables. 

$groupModels[$columnName] = $fieldModel; 

Thanks & regards,
Rajasekaran.

**Comment by vovpff** [comment]
@RajasekaranJeni try something like this instead code what you posted in this message https://github.com/YetiForceCompany/YetiForceCRM/discussions/15585#discussioncomment-1904963

```
$html .= '';
$groupModels = [];
// Labels array
$custLabels = ['qty' => 'YOUR LABEL FOR QTY',
'unit' => 'YOUR LABEL FOR UNIT',
];
foreach (['ItemNumber', 'Name', 'Quantity', 'UnitPrice', 'TotalPrice', 'Tax', 'GrossPrice'] as $fieldType) {
foreach ($inventory->getFieldsByType($fieldType) as $fieldModel) {
$columnName = $fieldModel->getColumnName();
if (!$fieldModel->isVisible()) {
continue;
}
$html .= "<th style="{$headerStyle}">";
// IF condition
if (array_key_exists($fieldModel->get('label'), $custLabels) {
$html .= $custLabel[$fieldModel->get('label')] . '';
} else {
$html .=  \App\Language::translate($fieldModel->get('label'), $this->textParser->moduleName) . '';
}
$groupModels[$columnName] = $fieldModel;
}
}
$html .= '</tr></thead>';`
```

This will work only with one language.
In case label the same as in array `$custLabels` this label will be used. Else standard mechanism of label translation will be used.

**Reply by Ilamparithi-Tamizhanangu** [reply]
Hi,

Thanks for the support. 

Please find the below table header section. I am trying to change the header for quantity.

$html .= '';
		$groupModels = [];
		// Labels array
		$custLabels = ['qty' => 'Quantity'];
		//'unit' => 'YOUR LABEL FOR UNIT',
		//];
		foreach (['ItemNumber', 'Name', 'Quantity', 'UnitPrice', 'TotalPrice', 'Tax', 'GrossPrice'] as $fieldType) {
		foreach ($inventory->getFieldsByType($fieldType) as $fieldModel) {
		$columnName = $fieldModel->getColumnName();
		if (!$fieldModel->isVisible()) {
		continue;
		}
		$html .= "<th style="{$headerStyle}">";
		// IF condition
		if (array_key_exists($fieldModel->get('label'), $custLabels) {
		$html .= $custLabel[$fieldModel->get('label')] . '';
		} else {
		$html .=  \App\Language::translate($fieldModel->get('label'), $this->textParser->moduleName) . '';
		}
		$groupModels[$columnName] = $fieldModel;
		}
		}
		$html .= '</tr></thead>';`

But I am getting the below error. 

![image](https://user-images.githubusercontent.com/96602176/148056740-6d892b37-4966-4573-9727-60440edb6ef1.png)

Please guide me to solve this error. 

Thanks & regards,
Rajasekaran.

**Comment by vovpff** [comment]
@RajasekaranJeni sorry, I write from smartphone

Try to relpace this string

```
$html .= "<th style="{$headerStyle}">";
```
 with
```
$html .= "<th style=\"{$headerStyle}\">";
```
or with 
```
$html .= '<th style="{' . $headerStyle .'}">';
```

**Reply by Ilamparithi-Tamizhanangu** [reply]
Hi,

Thanks for your efforts to solve our issues. 

Both the above codes showing the below error. 

![image](https://user-images.githubusercontent.com/96602176/148059220-adb4e129-1c91-4183-ae27-a8481122a687.png)

Please guide to me to solve this.

Thanks & regards,
Rajasekaran.

**Comment by vovpff** [comment]
@RajasekaranJeni try this

```
$html .= '';
$groupModels = [];
// Labels array
$custLabels = ['qty' => 'YOUR LABEL FOR QTY',
'unit' => 'YOUR LABEL FOR UNIT',
];
foreach (['ItemNumber', 'Name', 'Quantity', 'UnitPrice', 'TotalPrice', 'Tax', 'GrossPrice'] as $fieldType) {
foreach ($inventory->getFieldsByType($fieldType) as $fieldModel) {
$columnName = $fieldModel->getColumnName();
if (!$fieldModel->isVisible()) {
continue;
}
$html .= '<th style="{' . $headerStyle .'}">';
// IF condition
if (array_key_exists($fieldModel->get('label'), $custLabels) {
$html .= $custLabel[$fieldModel->get('label')] . '';
} else {
$html .= \App\Language::translate($fieldModel->get('label'), $this->textParser->moduleName) . '';
}
$groupModels[$columnName] = $fieldModel;
}
}
}
$html .= '</tr></thead>';
```

**Comment by Ilamparithi-Tamizhanangu** [comment]
Hi,

Please find the below code. 

$html .= '';
		$groupModels = [];
		// Labels array
		$custLabels = ['qty' => 'Quantity',
		'name' => 'Name',
		];
		foreach (['ItemNumber', 'Name', 'Quantity', 'UnitPrice', 'TotalPrice', 'Tax', 'GrossPrice'] as $fieldType) {
		foreach ($inventory->getFieldsByType($fieldType) as $fieldModel) {
		$columnName = $fieldModel->getColumnName();
		if (!$fieldModel->isVisible()) {
		continue;
		}
		$html .= '<th style="{' . $headerStyle .'}">';
		// IF condition
		if (array_key_exists($fieldModel->get('label'), $custLabels) {
		$html .= $custLabel[$fieldModel->get('label')] . '';
		} else {
		$html .= \App\Language::translate($fieldModel->get('label'), $this->textParser->moduleName) . '';
		}
		$groupModels[$columnName] = $fieldModel;
		}
		}
		}
		$html .= '</tr></thead>';`

Still i am getting the same error ( syntax error, unexpected ';').

I have one more doubt. We are closing the </tr> and </thead> tags in the last line. But we are not opening it. Does it create any error ? Kindly confirm. 

Thanks & regards,
Rajasekaran.

**Reply by vovpff** [reply]
Pls check my code for missed `) ` in if condition.

**Reply by Ilamparithi-Tamizhanangu** [reply]
Hi,

I tried to check the closing of the brackets in the complete code. But now we are getting different error. Please find the below complete code. 

`<?php

namespace App\TextParser;

/**
 * Products table short version class.
 *
 * @package TextParser
 *
 * @copyright YetiForce Sp. z o.o.
 * @license   YetiForce Public License 4.0 (licenses/LicenseEN.txt or yetiforce.com)
 * @author    Arkadiusz Sołek <a.solek@yetiforce.com>
 */
class ProductsTableMyVersion extends Base
{
	/** @var string Class name */
	public $name = 'LBL_PRODUCTS_TABLE_MY_VERSION';

	/** @var mixed Parser type */
	public $type = 'pdf';

	/**
	 * Process.
	 *
	 * @return string
	 */
	public function process()
	{
		$html = '';
		if (!$this->textParser->recordModel->getModule()->isInventory())
		{
			return $html;
		}
		$inventory = \Vtiger_Inventory_Model::getInstance($this->textParser->moduleName);
		$inventoryRows = $this->textParser->recordModel->getInventoryData();
		$firstRow = current($inventoryRows);
		$baseCurrency = \Vtiger_Util_Helper::getBaseCurrency();
		if ($inventory->isField('currency'))
		{
			if (!empty($firstRow) && null !== $firstRow['currency']) 
			{
				$currency = $firstRow['currency'];
			}
			else 
			{
				$currency = $baseCurrency['id'];
			}
			$currencyData = \App\Fields\Currency::getById($currency);
			$currencySymbol = $currencyData['currency_symbol'];
		}
		$headerStyle = 'font-size:9px;padding:0px 4px;text-align:center;';
		$bodyStyle = 'font-size:8px;border:1px solid #ddd;padding:0px 4px;';
		
		//custom code for the custom labels in Table headers
		
		$html .= '';
		$groupModels = [];
		// Custom Labels array
		$custLabels = ['qty' => 'Quantity','name' => 'Name',];
		foreach (['ItemNumber', 'Name', 'Quantity', 'UnitPrice', 'TotalPrice', 'Tax', 'GrossPrice'] as $fieldType)
		{
			foreach ($inventory->getFieldsByType($fieldType) as $fieldModel)
			{
				$columnName = $fieldModel->getColumnName();
				if (!$fieldModel->isVisible())
				{
					continue;
				}
				
				$html .= '<th style="{' . $headerStyle .'}">';
				// IF condition to select the custom or default labels
				if (array_key_exists($fieldModel->get('label'), $custLabels))
				{
					$html .= $custLabel[$fieldModel->get('label')] . '';
				}
				else
				{
					$html .= \App\Language::translate($fieldModel->get('label'), $this->textParser->moduleName) . '';
				}
				$groupModels[$columnName] = $fieldModel;
			}
		}
		$html .= '</tr></thead>';
		
		
		// Products sections

		if (!empty($groupModels))
		{
			$html .= '<tbody>';
			$counter = 0;
			foreach ($inventoryRows as $inventoryRow)
			{
				++$counter;
				$html .= '<tr class="row-' . $counter . '">';
				foreach ($groupModels as $fieldModel)
				{
					$columnName = $fieldModel->getColumnName();
					$typeName = $fieldModel->getType();
					$fieldStyle = $bodyStyle;
					if ('ItemNumber' === $typeName) 
					{
						$html .= "<td style=\"{$bodyStyle}font-weight:bold;\">" . $counter++ . '</td>';
					}
					elseif ('ean' === $columnName)
					{
						$code = $inventoryRow[$columnName];
						$html .= "<td class=\"col-type-barcode\" style=\"{$fieldStyle}font-weight:bold;text-align:center;\"><div data-barcode=\"EAN13\" data-code=\"{$code}\" data-size=\"1\" data-height=\"16\">{$code}</div></td>";
					}
					else
					{
						$itemValue = $inventoryRow[$columnName];
						if ('Name' === $typeName)
						{
							$fieldStyle = $bodyStyle . 'text-align:left;';
							$fieldValue = '<strong>' . $fieldModel->getDisplayValue($itemValue, $inventoryRow, true) . '</strong>';
							foreach ($inventory->getFieldsByType('Comment') as $commentField)
							{
								if ($commentField->isVisible() && ($value = $inventoryRow[$commentField->getColumnName()]) && $comment = $commentField->getDisplayValue($value, $inventoryRow, true)) 
								{
									$fieldValue .= '<br />' . $comment;
								}
							}
						} elseif (\in_array($typeName, ['GrossPrice', 'UnitPrice', 'TotalPrice', 'Tax']) && !empty($currencySymbol)) 
						{
							$fieldValue = \CurrencyField::appendCurrencySymbol($fieldModel->getDisplayValue($itemValue, $inventoryRow), $currencySymbol);
							$fieldStyle = $bodyStyle . 'text-align:right;white-space: nowrap;';
						}
						else 
						{
							$fieldValue = $fieldModel->getDisplayValue($itemValue, $inventoryRow, true);
						}
							$html .= "<td class=\"col-type-{$typeName}\" style=\"{$fieldStyle}\">" . $fieldValue . '</td>';
					}
					$html .= '</tr>';
				}
				$html .= '</tbody><tfoot><tr>';
				foreach ($groupModels as $fieldModel)
				{
					$headerStyle = 'font-size:7px;padding:0px 4px;text-align:center;';
					$html .= "<th class=\"col-type-{$typeName}\" style=\"{$headerStyle}\">";
					if ($fieldModel->isSummary())
					{
						$sum = 0;
						foreach ($inventoryRows as $inventoryRow) 
						{
							$sum += $inventoryRow[$fieldModel->getColumnName()];
						}
						$html .= \CurrencyField::appendCurrencySymbol(\CurrencyField::convertToUserFormat($sum, null, true), $currencySymbol);
					}
					$html .= '</th>';
				}
				$html .= '</tr></tfoot></table>';
			}
			return $html;
		}
	}
}`

This code gives the below error.
![image](https://user-images.githubusercontent.com/96602176/148207066-0a967aad-2bb7-4dd9-b9e3-a4662676e474.png)

Please help us to solve this error. 

Thanks & Regards,
Rajasekaran

**Reply by mariuszkrzaczkowski** [reply]
Attach the server configuration report

**Reply by Ilamparithi-Tamizhanangu** [reply]
Hi,

Please find the below report.

![ConfReport](https://user-images.githubusercontent.com/96602176/148210365-0ed42eeb-cfff-4fc6-a317-b645d33ceb9a.png)


Thanks & regards,
Rajasekaran.

**Reply by Ilamparithi-Tamizhanangu** [reply]
additional info: I can able to generate the PDF outputs for other formats. When i try to use the above codes to modify the products table header, i am getting the error.

**Comment by mariuszkrzaczkowski** [comment, marked as answer]
Look at this  https://github.com/YetiForceCompany/YetiForceCRM/commit/1749ffced87d3b8ea5373bf6b0ca08875390816

`
$(inventory : type=table columns=name,qty,price,total,tax,tax_percent,gross , href=no , labels=name xx,,,total yy,,,gross , width=40%,,,,,,400px )$
`

![2022-01-06_23-33-38](https://user-images.githubusercontent.com/10198654/148462071-03678a9e-5217-427f-b4e5-ae7d0551888a.png)
a

**Reply by Ilamparithi-Tamizhanangu** [reply]
hi,

I updated the TextParser.php file from above link to my yetiforce app folder and used the below code in the PDF template. 

$(inventory : type=table columns=seq,comment1,qty,price, href=no, labels=SNo,ProductDescription,QTY,UnitPrice, width=10%,50%,20%,30%, 600px)$

But i am unable to change the table headers or width in the PDF output. Please find the below output. 
![image](https://user-images.githubusercontent.com/96602176/148493820-32bd992e-23a6-4325-b9ad-fec08a07bc95.png)

Do i need to make any other changes. Please guide me on the same. 

Thanks.

**Reply by Ilamparithi-Tamizhanangu** [reply]
I am getting the similar output in the yetiforce developer edition in [this link](https://gitdeveloper.yetiforce.com/index.php?module=PDF&parent=Settings&view=List) also.  PDF template name is test 1.

please find the below output. 

![image](https://user-images.githubusercontent.com/96602176/148494257-6228c5ff-beab-4cab-b6ab-0191683188bf.png)

**Reply by mariuszkrzaczkowski** [reply]
it works see, watch out because every sign matters
```
$(inventory : type=table columns=seq,comment1,qty,price, href=no, labels=SNo,ProductDescription,QTY,UnitPrice, width=10%,50%,20%,30%, 600px)$

$(inventory : type=table columns=seq,comment1,qty,price , href=no , labels=SNo,ProductDescription,QTY,UnitPrice , width=10%,50%,20%,30%,600px)$
```
